### PR TITLE
work in progress: Custom json encoder/decoder

### PR DESCRIPTION
**Pull request by linoor**
Cf. #99 

### Successful PR Checklist:
- [ ] Tests
- [ ] Documentation (optionally: [run spell checking](https://github.com/peopledoc/procrastinate/blob/master/CONTRIBUTING.rst#build-the-documentation))
- [ ] Had a good time contributing? (if not, feel free to give some feedback)

### CODE REVIEW

**Comment by ewjoachim** [comment]
This looks like it's going the right way.

Small nudge: have a look at
```python
# always consider that dicts are JSON (not hstore)
psycopg2.extensions.register_adapter(dict, JsonAdapter)
```

I believe we might be able to get rid of `wrap_json` altogether (see https://github.com/peopledoc/procrastinate/pull/96#discussion_r339553296)

If you need anything, let me know :)

File: procrastinate/json_adapter.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class JsonAdapter(psycopg2.extras.Json):
+    """
+    Convert a given dict to JSON, while handling conversion of some
+    otherwise non JSON serialisable objects.
+    """
+
+    def encoder(self, obj) -> str:
+        if isinstance(obj, (date, datetime)):
+            # returns the date or datetime in ISO format
+            return str(obj)
```
Comment (hsmett) [inline procrastinate/json_adapter.py]: Do we want the str representation or the iso format ?
```
In: utc_now = pytz.utc.localize(datetime.datetime.utcnow())                                                                                                                                                   

In: str(utc_now)                                                                                                                                                                                              
Out: '2019-11-18 09:15:58.038718+00:00'

In : utc_now.isoformat()                                                                                                                                                                                       
Out: '2019-11-18T09:15:58.038718+00:00'
```

**Comment by ewjoachim** [comment]
Yes ! So forget my comment above (https://github.com/peopledoc/procrastinate/pull/108#issuecomment-554628468) and keep doing what you've been doing, it appears you're already on the best path :clap:

**Comment by elemoine** [comment]
For everyone's information, I've started working on the same functionality, but using a different approach.

I see three problems with this PR:

a) It does this:

```py
psycopg2.extensions.register_adapter(dict, JsonAdapter)
```

which, as @k4nar said it already, may have an impact on the application.

b) The JsonAdapter is part of procrastinate, while we want the user to be able to provide custom encoders/decoders dealing with the specific types he or she might have.

c) This PR's approach only deals with the encoding part, and the decoding part is required for the worker side.

I will soon open a new PR describing my approach.

**Comment by elemoine** [comment]
Superseded by #131.

**Pull request closed without merging**